modules/ipc/include/hephaestus/ipc/zenoh/action_server.h

```diff
@@ -0,0 +1,250 @@
+/// Upon completion a result is sent back to the client.
+/// Differently from classic request/response servers, action servers are asynchronous and non-blocking.
+/// Action servers also provide functionalities for the user to send status updates to the client and are
+/// interruptible.
+/// To instantiate an ActionServer the user needs to provide two callbacks:
+/// - RequestCallback
+///   - Takes in input the request and need to decide if the request is valid and if it can be served.
+///   - The callback should not block and return immediately,
+/// - ExecuteCallback
+///   - This is the function that does the actual work and eventually return the final response to the client.
+///   - The execute callback is executed in a dedicated thread.
+///   - The function begin execution as soon as the request is accepted.
+///   - The function has access to a Publisher to send status updates to the client.
+///     - The frequency of the updates is decided by the user; updates are not mandatory
+///   - The function can be interrupted by the client by setting the stop_requested flag to true.
+///     - The function should check this flag periodically and return if it is set.
+/// NOTES:
+/// - If an action server is already serving a request it will not accept new ones.
+/// - RequestT needs to be copyable and one copy will be made of it.
+/// ---------------------------------------------------------------------------------------------------------
+/// Implementation details:
+/// - ActionServer contains a `Service` to receive the requests and a `MessageQueueConsumer` to execute them.
+/// - When a new request is accepted and started a new `Publisher` is created to send status updates.
+/// - The publisher is passed to `execute_cb`, which the user can use to publish status update of the
+///   execution.
+///   - The client when calling the server will create a temporary `Subscriber` to receive the updates.
+/// - When `execute_cb` finishes the final response is sent to the client via a `Service` created by the
+///   caller.
+/// NOTE: The reason why we can only process one request at a time is that with the current implementation the
+/// response `Service` and the update `Publisher` uses a topic name which just depends on the input topic
+/// name. This means that if we have multiple requests we will have multiple `Publisher` and `Subscriber` with
+/// the same topic name, which will cause conflicts. This could be solved by adding a unique identifier to the
+/// topic name, but this is not implemented yet.
+template <typename RequestT, typename StatusT, typename ReplyT>
+class ActionServer {
+public:
+  using RequestCallback = std::function<ActionServerRequestStatus(const RequestT&)>;
+  using ExecuteCallback =
+      std::function<ReplyT(const RequestT&, Publisher<StatusT>&, std::atomic_bool& stop_requested)>;
+
+  ActionServer(SessionPtr session, TopicConfig topic_config, RequestCallback&& request_cb,
```

Comment:
we should probably also have a configurable timeout on the request server side?

Reply:
how do you implement it? We go back to killing the user code